lib/ecto/type.ex

```diff
@@ -101,10 +101,10 @@ defmodule Ecto.Type do
   @typep base      :: :integer | :float | :boolean | :string | :map |
                       :binary | :decimal | :id | :binary_id |
                       :datetime | :date | :time | :any
-  @typep composite :: {:array, base} | {:embed, Ecto.Embedded.t} | {:in, base}
+  @typep composite :: {:array, base} | {:map, base} | {:embed, Ecto.Embedded.t} | {:in, base}
```

Comment:
we can have arbitrary complex types e.g.: `{:map, {:array, {:map, :map}}}`, I'm not sure how to represent them in typespec though